File: app/src/main/java/com/dialogapp/dialog/ui/common/RequestViewModel.java
package com.dialogapp.dialog.ui.common;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.arch.lifecycle.ViewModel;

import com.dialogapp.dialog.repository.PostRequestManager;
import com.dialogapp.dialog.util.Event;
import com.dialogapp.dialog.util.Objects;
import com.dialogapp.dialog.util.Resource;

import javax.inject.Inject;

public class RequestViewModel extends ViewModel {
    private final MutableLiveData<User> user = new MutableLiveData<>();
    private LiveData<Event<Resource<Boolean>>> responseFollow;

    private final MutableLiveData<Favorite> favorite = new MutableLiveData<>();
    private LiveData<Event<Resource<Boolean>>> responseFavorite;

    private final MutableLiveData<Reply> reply = new MutableLiveData<>();
    private LiveData<Event<Resource<Boolean>>> responseReply;

    @Inject
    public RequestViewModel(PostRequestManager postRequestManager) {
        responseFollow = Transformations.switchMap(user,
                input -> postRequestManager.followUser(input.username, input.shouldFollow));

        responseFavorite = Transformations.switchMap(favorite,
                input -> postRequestManager.sendFavoriteRequest(input.id, input.shouldFavorite));

        responseReply = Transformations.switchMap(reply, input -> postRequestManager.sendReply(input.id, input.text));
    }

    public void setFollowState(String username, boolean shouldFollow) {
        this.user.setValue(new User(username, shouldFollow));
    }

    public LiveData<Event<Resource<Boolean>>> getResponseFollow() {
        return responseFollow;
    }

    public void setFavoriteState(String id, boolean shouldFavorite) {
        this.favorite.setValue(new Favorite(id, shouldFavorite));
    }

    public LiveData<Event<Resource<Boolean>>> getResponseFavorite() {
        return responseFavorite;
    }

    public void sendReply(String id, String text) {
        Reply update = new Reply(id, text);
        if (Objects.equals(this.reply.getValue(), update))
            return;
        this.reply.setValue(new Reply(id, text));
    }

    public LiveData<Event<Resource<Boolean>>> getResponseReply() {
        return responseReply;
    }

    public void retryReply() {
        Reply current = this.reply.getValue();
        if (current != null && !current.isEmpty())
            this.reply.setValue(current);
    }

    public static class User {
        public final String username;
        public final boolean shouldFollow;

        public User(String username, boolean shouldFollow) {
            this.username = username;
            this.shouldFollow = shouldFollow;
        }
    }

    public static class Favorite {
        public final String id;
        public final boolean shouldFavorite;

        public Favorite(String id, boolean shouldFavorite) {
            this.id = id;
            this.shouldFavorite = shouldFavorite;
        }
    }

    private class Reply {
        public final String id;
        public final String text;

        public Reply(String id, String text) {
            this.id = id;
            this.text = text;
        }

        boolean isEmpty() {
            return id == null || text == null || id.length() == 0 || text.length() == 0;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (o == null || getClass() != o.getClass()) {
                return false;
            }

            Reply reply = (Reply) o;

            if (id != null ? !id.equals(reply.id) : reply.id != null) {
                return false;
            }
            return text != null ? text.equals(reply.text) : reply.text == null;
        }

        @Override
        public int hashCode() {
            int result = id != null ? id.hashCode() : 0;
            result = 31 * result + (text != null ? text.hashCode() : 0);
            return result;
        }
    }
}

File: app/src/main/java/com/dialogapp/dialog/ui/common/RequestViewModel.kt
package com.dialogapp.dialog.ui.common

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import androidx.work.ExistingWorkPolicy
import androidx.work.OneTimeWorkRequest
import androidx.work.WorkInfo
import androidx.work.WorkManager
import com.dialogapp.dialog.util.Event
import com.dialogapp.dialog.workers.*

class RequestViewModel : ViewModel() {

    var title: String? = null
        set(value) {
            field = if (value != null && value.isBlank())
                null
            else
                value
        }

    private val _replyTag = MutableLiveData<String>()
    private val _newPostTag = MutableLiveData<String>()
    private val _deleteTag = MutableLiveData<String>()

    private var workManager: WorkManager = WorkManager.getInstance()
    private var replyWorkInfo: LiveData<List<WorkInfo>>? = Transformations
            .switchMap(_replyTag) { tag ->
                workManager.getWorkInfosByTagLiveData(tag)
            }
    private var newPostWorkInfo: LiveData<List<WorkInfo>>? = Transformations
            .switchMap(_newPostTag) { tag ->
                workManager.getWorkInfosByTagLiveData(tag)
            }
    private var deleteWorkInfo: LiveData<List<WorkInfo>>? = Transformations
            .switchMap(_deleteTag) { tag ->
                workManager.getWorkInfosByTagLiveData(tag)
            }
    private lateinit var draftReply: Pair<String, String>
    private lateinit var draftPost: Pair<String?, String>
    lateinit var replyEvent: Event<Boolean>
        private set
    lateinit var postEvent: Event<Boolean>
        private set
    lateinit var deleteEvent: Event<Boolean>
        private set

    fun sendFavoriteRequest(postId: String?, belongsToEndpoint: String?) {
        val tag = "FAV_$postId"
        val favoriteRequest = OneTimeWorkRequest.Builder(FavoriteWorker::class.java)
                .setInputData(FavoriteWorker.createInputData(postId, belongsToEndpoint))
                .addTag(tag)
                .build()
        workManager.enqueueUniqueWork(tag, ExistingWorkPolicy.KEEP, favoriteRequest)
    }

    fun sendDeleteRequest(postId: String) {
        val tag = "DEL_$postId"
        val deletePostRequest = OneTimeWorkRequest.Builder(DeletePostWorker::class.java)
                .setInputData(DeletePostWorker.createInputData(postId))
                .addTag(tag)
                .build()
        workManager.enqueueUniqueWork(tag, ExistingWorkPolicy.KEEP, deletePostRequest)
        _deleteTag.value = tag
        deleteEvent = Event(true)
    }

    fun sendFollowRequest(username: String, isFollowing: Boolean) {
        val tag = "FOL_$username"
        val followRequest = OneTimeWorkRequest.Builder(FollowWorker::class.java)
                .setInputData(FollowWorker.createInputData(username, isFollowing))
                .addTag(tag)
                .build()
        WorkManager.getInstance().enqueueUniqueWork(tag, ExistingWorkPolicy.KEEP, followRequest)
    }

    fun sendReply(postId: String, content: String) {
        draftReply = Pair(postId, content)
        val tag = "REP_$postId"
        val replyRequest = OneTimeWorkRequest.Builder(ReplyWorker::class.java)
                .setInputData(ReplyWorker.createInputData(postId, content))
                .addTag(tag)
                .build()
        workManager.enqueueUniqueWork(tag, ExistingWorkPolicy.KEEP, replyRequest)
        _replyTag.value = tag
        replyEvent = Event(true)
    }

    fun sendPost(title: String? = this.title, content: String) {
        draftPost = Pair(title, content)
        val tag = "NEW"
        val newPostRequest = OneTimeWorkRequest.Builder(NewPostWorker::class.java)
                .setInputData(NewPostWorker.createInputData(title, content))
                .addTag(tag)
                .build()
        workManager.enqueueUniqueWork(tag, ExistingWorkPolicy.KEEP, newPostRequest)
        _newPostTag.value = tag
        postEvent = Event(true)
    }

    fun getReplyWorkInfo(): LiveData<List<WorkInfo>>? {
        return replyWorkInfo
    }

    fun getNewPostWorkInfo(): LiveData<List<WorkInfo>>? {
        return newPostWorkInfo
    }

    fun getDeleteWorkInfo(): LiveData<List<WorkInfo>>? {
        return deleteWorkInfo
    }

    fun retryReply() {
        sendReply(draftReply.first, draftReply.second)
    }

    fun retryPost() {
        sendPost(draftPost.first, draftPost.second)
    }
}